src/cost_estimates/fee_medians.rs

```diff
@@ -0,0 +1,359 @@
+///    1f, the minimum. Ignoring the amount of empty space leads to over-estimates because it
+///    ignores the fact that there was still space in the block.
+pub struct WeightedMedianFeeRateEstimator<M: CostMetric> {
+    db: Connection,
+    /// We only look back `window_size` fee rates when averaging past estimates.
+    window_size: u32,
+    /// The weight of a "full block" in abstract scalar cost units. This is the weight of
+    /// a block that is filled *one single* dimension.
+    full_block_weight: u64,
+    /// Use this cost metric in fee rate calculations.
+    metric: M,
+}
+
+/// Convenience struct for passing around this pair.
+#[derive(Debug)]
+pub struct FeeRateAndWeight {
+    pub fee_rate: f64,
+    pub weight: u64,
+}
+
+impl<M: CostMetric> WeightedMedianFeeRateEstimator<M> {
+    /// Open a fee rate estimator at the given db path. Creates if not existent.
+    pub fn open(p: &Path, metric: M, window_size: u32) -> Result<Self, SqliteError> {
+        let mut db = sqlite_open(
+            p,
+            rusqlite::OpenFlags::SQLITE_OPEN_CREATE | rusqlite::OpenFlags::SQLITE_OPEN_READ_WRITE,
+            false,
+        )?;
+
+        // check if the db needs to be instantiated regardless of whether or not
+        //  it was newly created: the db itself may be shared with other fee estimators,
+        //  which would not have created the necessary table for this estimator.
+        let tx = tx_begin_immediate_sqlite(&mut db)?;
+        Self::instantiate_db(&tx)?;
+        tx.commit()?;
+
+        Ok(Self {
+            db,
+            metric,
+            window_size,
+            full_block_weight: PROPORTION_RESOLUTION,
+        })
+    }
+
+    /// Check if the SQL database was already created. Necessary to avoid races if
+    ///  different threads open an estimator at the same time.
+    fn db_already_instantiated(tx: &SqlTransaction) -> Result<bool, SqliteError> {
+        table_exists(tx, "median_fee_estimator")
+    }
+
+    fn instantiate_db(tx: &SqlTransaction) -> Result<(), SqliteError> {
+        if !Self::db_already_instantiated(tx)? {
+            tx.execute(CREATE_TABLE, rusqlite::NO_PARAMS)?;
+        }
+
+        Ok(())
+    }
+
+    fn get_rate_estimates_from_sql(
+        conn: &Connection,
+        window_size: u32,
+    ) -> Result<FeeRateEstimate, EstimatorError> {
+        let sql =
+            "SELECT high, middle, low FROM median_fee_estimator ORDER BY measure_key DESC LIMIT ?";
+        let mut stmt = conn.prepare(sql).expect("SQLite failure");
+
+        // shuttle high, low, middle estimates into these lists, and then sort and find median.
+        let mut highs = Vec::with_capacity(window_size as usize);
+        let mut mids = Vec::with_capacity(window_size as usize);
+        let mut lows = Vec::with_capacity(window_size as usize);
+        let results = stmt
+            .query_and_then::<_, SqliteError, _, _>(&[window_size], |row| {
```

Comment:
In rusqlite, you can get columns by name instead of index.  It's generally safer to do this, since it's far less brittle.